designer/client/src/data/condition/updateConditions.ts

```diff
@@ -0,0 +1,70 @@
+import {
+  hasConditionField,
+  hasConditionGroup,
+  type ConditionData,
+  type ConditionWrapper,
+  type FormDefinition
+} from '@defra/forms-model'
+
+import { getFields } from '~/src/data/component/fields.js'
+
+/**
+ * Update conditions to keep display text in sync with the component and section titles
+ */
+export function updateConditions(data: FormDefinition) {
+  const fields = getFields(data)
+
+  // Check if condition field display text is out of date
+  const checkCondition = (condition: ConditionData) => {
+    const fieldName = condition.field.name
+    const field = fields.find((input) => input.name === fieldName)
+
+    return field && condition.field.display !== field.label ? field : undefined
+  }
+
+  // Determine if any of the conditions need fixing
+  const hasConditionsChanged = data.conditions.some((wrapper) => {
+    return wrapper.value.conditions.some((condition) => {
+      if (hasConditionField(condition)) {
+        return checkCondition(condition)
+      } else if (hasConditionGroup(condition)) {
+        condition.conditions
+          .filter((condition) => hasConditionField(condition))
+          .forEach(checkCondition)
+      }
+
+      return false
+    })
+  })
+
+  if (!hasConditionsChanged) {
+    return data
+  }
+
+  // Copy form definition
+  const definition = structuredClone(data)
+  const { conditions } = definition
+
+  // Fix up condition field display text
+  const fixUpCondition = (condition: ConditionData) => {
+    const field = checkCondition(condition)
+
+    if (field) {
+      condition.field.display = field.label
+    }
+  }
+
+  conditions.forEach((condition: ConditionWrapper) => {
```

Comment:
I thought we had an ESLint rule for this but we can remove types when they're already known

```suggestion
  conditions.forEach((condition) => {
```